Stats: Fixes simplification bug with Stochastic Process

0sidharth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
#20078 

#### Brief description of what is fixed or changed
Before fix:
```
>>> print((2*W(t)).simplify())
>>> AttributeError: 'function' object has no attribute 'is_Rational'
```

After fix:
```
>>> print((2*W(t)).simplify())
>>> 2*W(t)
```

#### Other comments

Changes originally [proposed](https://gist.github.com/czgdp1807/879223647496d9cef0b246272731c3db) by @czgdp1807
A minor change was made in pspace property of `RandomIndexedSymbol`, to allow for calling `RandomIndexedSymbol` with one argument.

This is my first time contributing to sympy, so please guide me if I have made any mistakes. Thank you!

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
    * Fixed simplification bug in Stochastic Processes

<!-- END RELEASE NOTES -->

theHamsta committed:
Replace hasattr(... '__iter__') by check with `is_sequence`

oscarbenjamin committed:
maint(test): add optional dependency tests to actions

FFroehlich committed:
fix cxx printing for Min/Max

FFroehlich committed:
remove trailing whitespace

AlexCQY committed:
update method in ode.py

AlexCQY committed:
add test for issue #13060

FFroehlich committed:
add tests

FFroehlich committed:
remove trailing whitespaces

AlexCQY committed:
fix code quality issue

FFroehlich committed:
fixup

oscarbenjamin committed:
maint(testing): add test_optional_dependencies script for actions

oscarbenjamin committed:
maint(testing): use TestsFailedError instead of Exception

oscarbenjamin committed:
maint(testing): run sphinx build on actions

0sidharth committed:
aims to fix simplification bug (#20078) with Wiener Process

czgdp1807 commented on sympy/stats/tests/test_stochastic_process.py:
```diff
@@ -643,6 +643,9 @@ def test_WienerProcess():
     assert E(X(t) + x*E(X(3))) == 0
 
+    # test issue 20078
+    assert str((2*X(t)).simplify()) == '2*X(t)'
```
String comparisons should be avoided.

czgdp1807 replied:
Add more tests like, `simplify(X(t) + X(t)) == 2*X(t)` to verify that the simplification actually works.

czgdp1807 commented on sympy/stats/tests/test_stochastic_process.py:
```diff
@@ -652,7 +655,7 @@ def test_GammaProcess_symbolic():
     assert isinstance(X(t), RandomIndexedSymbol)
     assert X.state_space == Interval(0, oo)
-    assert X.distribution(X(t)) == GammaDistribution(g*t, 1/l)
+    assert X.distribution(t) == GammaDistribution(g*t, 1/l)
```
This isn't private API. Add a deprecation warning in `distribution`.
Possible add a function, like, `_deprecation_warn_distribution(self)` in the root level class and write the deprecation warning code inside it. Then call that function in the distribution method of each subclass.

czgdp1807 commented:
This doesn't fully fixes the issues. The problem with derivative requires a different fix where we need to define the diff method for each type of stochastic process. Should be done in a different PR.

0sidharth commented:
I have added the deprecations warnings, and I have also added back the old implementations of all `distribution` functions changed so that this change is not a breaking change. But this would require that sometime in the future that this old implementation is removed for good, so I am not sure whether this is the right way to go (I see `last_supported_version` in `SymPyDeprecationWarning` so I am guessing that this is the way to handle breaking changes), can you confirm this?

I have also added a few tests that made sense to me if I am missing any or they don't make sense, do tell.

czgdp1807 commented:
Is the following issue related,
```python
IPython console for SymPy 1.8.dev (Python 3.6.9-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/dev


In [1]: from sympy.stats import *

In [2]: T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])

In [3]: Y = DiscreteMarkovChain("Y", [0, 1, 2], T)

In [4]: Y[2] + Y[2]
Out[4]: 2⋅Y[2]

In [5]: simplify(Y[2]**2 + Y[2]**3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/sympy_project/sympy/sympy/core/cache.py in wrapper(*args, **kwargs)
     71             try:
---> 72                 retval = cfunc(*args, **kwargs)
     73             except TypeError:

~/sympy_project/sympy/sympy/core/operations.py in __new__(cls, evaluate, _sympify, *args)
     84 
---> 85         c_part, nc_part, order_symbols = cls.flatten(args)
     86         is_commutative = not nc_part

~/sympy_project/sympy/sympy/core/add.py in flatten(cls, seq)
    273         # order args canonically
--> 274         _addsort(newseq)
    275 

~/sympy_project/sympy/sympy/core/add.py in _addsort(args)
     18     # in-place sorting of args
---> 19     args.sort(key=_args_sortkey)
     20 

~/sympy_project/sympy/sympy/core/basic.py in compare(self, other)
    223             if isinstance(l, Basic):
--> 224                 c = l.compare(r)
    225             else:

~/sympy_project/sympy/sympy/core/basic.py in compare(self, other)
    223             if isinstance(l, Basic):
--> 224                 c = l.compare(r)
    225             else:

~/sympy_project/sympy/sympy/core/basic.py in compare(self, other)
    225             else:
--> 226                 c = (l > r) - (l < r)
    227             if c:

TypeError: '>' not supported between instances of 'NoneType' and 'NoneType'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-5-469ca52a2f87> in <module>
----> 1 simplify(Y[2]**2 + Y[2]**3)

~/sympy_project/sympy/sympy/core/decorators.py in _func(self, other)
    265             if not isinstance(other, expectedcls):
    266                 return retval
--> 267             return func(self, other)
    268 
    269         return _func

~/sympy_project/sympy/sympy/core/decorators.py in binary_op_wrapper(self, other)
    135                     if f is not None:
    136                         return f(self)
--> 137             return func(self, other)
    138         return binary_op_wrapper
    139     return priority_decorator

~/sympy_project/sympy/sympy/core/expr.py in __add__(self, other)
    195     @call_highest_priority('__radd__')
    196     def __add__(self, other):
--> 197         return Add(self, other)
    198 
    199     @sympify_return([('other', 'Expr')], NotImplemented)

~/sympy_project/sympy/sympy/core/cache.py in wrapper(*args, **kwargs)
     72                 retval = cfunc(*args, **kwargs)
     73             except TypeError:
---> 74                 retval = func(*args, **kwargs)
     75             return retval
     76 

~/sympy_project/sympy/sympy/core/operations.py in __new__(cls, evaluate, _sympify, *args)
     83             return args[0]
     84 
---> 85         c_part, nc_part, order_symbols = cls.flatten(args)
     86         is_commutative = not nc_part
     87         obj = cls._from_args(c_part + nc_part, is_commutative)

~/sympy_project/sympy/sympy/core/add.py in flatten(cls, seq)
    272 
    273         # order args canonically
--> 274         _addsort(newseq)
    275 
    276         # current code expects coeff to be first

~/sympy_project/sympy/sympy/core/add.py in _addsort(args)
     17 def _addsort(args):
     18     # in-place sorting of args
---> 19     args.sort(key=_args_sortkey)
     20 
     21 

~/sympy_project/sympy/sympy/core/basic.py in compare(self, other)
    222             r = Basic(*r) if isinstance(r, frozenset) else r
    223             if isinstance(l, Basic):
--> 224                 c = l.compare(r)
    225             else:
    226                 c = (l > r) - (l < r)

~/sympy_project/sympy/sympy/core/basic.py in compare(self, other)
    222             r = Basic(*r) if isinstance(r, frozenset) else r
    223             if isinstance(l, Basic):
--> 224                 c = l.compare(r)
    225             else:
    226                 c = (l > r) - (l < r)

~/sympy_project/sympy/sympy/core/basic.py in compare(self, other)
    224                 c = l.compare(r)
    225             else:
--> 226                 c = (l > r) - (l < r)
    227             if c:
    228                 return c

TypeError: '>' not supported between instances of 'NoneType' and 'NoneType'
```

0sidharth commented:
The error messages are the same as the ones for the Wiener Process, so I believe this might be related. Looking into it.

0sidharth commented:
The error was in this line

https://github.com/0sidharth/sympy/blob/d1cd9ad36d26418f1e405fa00b9d5facafb6a963/sympy/stats/stochastic_process_types.py#L273

`DiscreteMarkovChain` and `ContinuousMarkovChain` do not have `distribution` function, instead they have `stationary_distribution` and `limiting_distribution` respectively. I have changed the code (still testing before pushing) to check if the Process is a `MarkovChain` or not, and accordingly call the distribution function.

czgdp1807 commented on sympy/stats/stochastic_process_types.py:
```diff
@@ -274,7 +293,10 @@ def __call__(self, time):
         if not time.is_symbol and time not in self.index_set:
             raise IndexError("%s is not in the index set of %s"%(time, self.symbol))
         func_obj = Function(self.symbol)(time)
-        pspace_obj = StochasticPSpace(self.symbol, self, self.distribution)
```
What if we would have just called `self.distribution()`.

0sidharth replied:
We still need to return `RandomIndexedSymbol`, so the `pspace_obj` needs to be created. Also, there is no implementation of `distribution()` method for `MarkovChain`, as I explained in my previous comment. So the error won't be resolved for `MarkovChain` Processes unless we pass their specific distribution methods as the 3rd argument to `RandomIndexedSymbol` (The 3rd argument would just remain `None` otherwise, which was causing the error)

So I think it is necessary to proceed this way, or did you have any other method in mind?

czgdp1807 replied:
What about returning an object of abstract `Distribution` class?
```diff
diff --git a/sympy/core/tests/test_args.py b/sympy/core/tests/test_args.py
index d33b261750..6f9c984123 100644
--- a/sympy/core/tests/test_args.py
+++ b/sympy/core/tests/test_args.py
@@ -1050,6 +1050,10 @@ def test_sympy__stats__crv__SingleContinuousPSpace():
     from sympy.stats.crv import SingleContinuousPSpace
     assert _test_args(SingleContinuousPSpace(x, nd))
 
+@SKIP("abstract class")
+def test_sympy__stats__rv__Distribution():
+    pass
+
 @SKIP("abstract class")
 def test_sympy__stats__crv__SingleContinuousDistribution():
     pass
diff --git a/sympy/stats/compound_rv.py b/sympy/stats/compound_rv.py
index da3fe28250..b2fad26abf 100644
--- a/sympy/stats/compound_rv.py
+++ b/sympy/stats/compound_rv.py
@@ -1,6 +1,6 @@
 from sympy import Basic, Sum, Dummy, Lambda, Integral
 from sympy.stats.rv import (NamedArgsMixin, random_symbols, _symbol_converter,
-                        PSpace, RandomSymbol, is_random)
+                        PSpace, RandomSymbol, is_random, Distribution)
 from sympy.stats.crv import ContinuousDistribution, SingleContinuousPSpace
 from sympy.stats.drv import DiscreteDistribution, SingleDiscretePSpace
 from sympy.stats.frv import SingleFiniteDistribution, SingleFinitePSpace
@@ -120,7 +120,7 @@ def conditional_space(self, condition, *, compound_evaluate=True, **kwargs):
         return new_pspace.conditional_space(condition)
 
 
-class CompoundDistribution(Basic, NamedArgsMixin):
+class CompoundDistribution(Distribution, NamedArgsMixin):
     """
     Class for Compound Distributions.
 
diff --git a/sympy/stats/crv.py b/sympy/stats/crv.py
index 07e8440dd8..9a4faff4ae 100644
--- a/sympy/stats/crv.py
+++ b/sympy/stats/crv.py
@@ -21,7 +21,7 @@
 from sympy.core.sympify import _sympify
 from sympy.external import import_module
 from sympy.stats.rv import (RandomDomain, SingleDomain, ConditionalDomain, is_random,
-        ProductDomain, PSpace, SinglePSpace, random_symbols, NamedArgsMixin)
+        ProductDomain, PSpace, SinglePSpace, random_symbols, NamedArgsMixin, Distribution)
 
 
 class ContinuousDomain(RandomDomain):
@@ -140,7 +140,7 @@ def set(self):
                 "Set of Conditional Domain not Implemented")
 
 
-class ContinuousDistribution(Basic):
+class ContinuousDistribution(Distribution):
     def __call__(self, *args):
         return self.pdf(*args)
 
diff --git a/sympy/stats/drv.py b/sympy/stats/drv.py
index a037386503..e5a1f09750 100644
--- a/sympy/stats/drv.py
+++ b/sympy/stats/drv.py
@@ -6,7 +6,7 @@
 from sympy.stats.crv import reduce_rational_inequalities_wrap
 from sympy.stats.rv import (NamedArgsMixin, SinglePSpace, SingleDomain,
                             random_symbols, PSpace, ConditionalDomain, RandomDomain,
-                            ProductDomain)
+                            ProductDomain, Distribution)
 from sympy.stats.symbolic_probability import Probability
 from sympy.sets.fancysets import Range, FiniteSet
 from sympy.sets.sets import Union
@@ -16,7 +16,7 @@
 from sympy.external import import_module
 
 
-class DiscreteDistribution(Basic):
+class DiscreteDistribution(Distribution):
     def __call__(self, *args):
         return self.pdf(*args)
 
diff --git a/sympy/stats/frv.py b/sympy/stats/frv.py
index ef24efaf45..98f2cc8499 100644
--- a/sympy/stats/frv.py
+++ b/sympy/stats/frv.py
@@ -19,7 +19,7 @@
 from sympy.sets.sets import FiniteSet
 from sympy.stats.rv import (RandomDomain, ProductDomain, ConditionalDomain,
                             PSpace, IndependentProductPSpace, SinglePSpace, random_symbols,
-                            sumsets, rv_subs, NamedArgsMixin, Density)
+                            sumsets, rv_subs, NamedArgsMixin, Density, Distribution)
 from sympy.external import import_module
 
 
@@ -181,7 +181,7 @@ def set(self):
     def as_boolean(self):
         return FiniteDomain.as_boolean(self)
 
-class SingleFiniteDistribution(Basic, NamedArgsMixin):
+class SingleFiniteDistribution(Distribution, NamedArgsMixin):
     def __new__(cls, *args):
         args = list(map(sympify, args))
         return Basic.__new__(cls, *args)
diff --git a/sympy/stats/joint_rv.py b/sympy/stats/joint_rv.py
index b2a6bc87d0..abbbcaa991 100644
--- a/sympy/stats/joint_rv.py
+++ b/sympy/stats/joint_rv.py
@@ -20,8 +20,9 @@
 from sympy.matrices import ImmutableMatrix, matrix2numpy, list2numpy
 from sympy.stats.crv import SingleContinuousDistribution, SingleContinuousPSpace
 from sympy.stats.drv import SingleDiscreteDistribution, SingleDiscretePSpace
-from sympy.stats.rv import (ProductPSpace, NamedArgsMixin,
-                            ProductDomain, RandomSymbol, random_symbols, SingleDomain, _symbol_converter)
+from sympy.stats.rv import (ProductPSpace, NamedArgsMixin, Distribution,
+                            ProductDomain, RandomSymbol, random_symbols,
+                            SingleDomain, _symbol_converter)
 from sympy.utilities.misc import filldedent
 from sympy.external import import_module
 
@@ -232,7 +233,7 @@ def _sample_pymc3(cls, dist, size):
     'numpy': SampleJointNumpy
 }
 
-class JointDistribution(Basic, NamedArgsMixin):
+class JointDistribution(Distribution, NamedArgsMixin):
     """
     Represented by the random variables part of the joint distribution.
     Contains methods for PDF, CDF, sampling, marginal densities, etc.
@@ -306,7 +307,7 @@ def __getitem__(self, key):
 
 
 
-class MarginalDistribution(Basic):
+class MarginalDistribution(Distribution):
     """
     Represents the marginal distribution of a joint probability space.
 
diff --git a/sympy/stats/matrix_distributions.py b/sympy/stats/matrix_distributions.py
index 4307b2dc64..8f945ed997 100644
--- a/sympy/stats/matrix_distributions.py
+++ b/sympy/stats/matrix_distributions.py
@@ -4,7 +4,7 @@
                             MatrixSymbol, MatrixBase, Transpose, MatrixSet,
                             matrix2numpy)
 from sympy.stats.rv import (_value_check, RandomMatrixSymbol, NamedArgsMixin, PSpace,
-                            _symbol_converter, MatrixDomain)
+                            _symbol_converter, MatrixDomain, Distribution)
 from sympy.external import import_module
 
 
@@ -153,7 +153,7 @@ def _sample_pymc3(cls, dist, size):
 #-------------------------Matrix Distribution----------------------------------#
 ################################################################################
 
-class MatrixDistribution(Basic, NamedArgsMixin):
+class MatrixDistribution(Distribution, NamedArgsMixin):
     """
     Abstract class for Matrix Distribution
     """
diff --git a/sympy/stats/rv.py b/sympy/stats/rv.py
index d869d216a6..f4a02f640f 100644
--- a/sympy/stats/rv.py
+++ b/sympy/stats/rv.py
@@ -349,6 +349,9 @@ def __new__(cls, symbol, n, m, pspace=None):
     symbol = property(lambda self: self.args[0])
     pspace = property(lambda self: self.args[3])
 
+class Distribution(Basic):
+    pass
+
 class ProductPSpace(PSpace):
     """
     Abstract class for representing probability spaces with multiple random
diff --git a/sympy/stats/stochastic_process_types.py b/sympy/stats/stochastic_process_types.py
index d1d05fe565..175f43740a 100644
--- a/sympy/stats/stochastic_process_types.py
+++ b/sympy/stats/stochastic_process_types.py
@@ -20,7 +20,7 @@
 from sympy.stats.joint_rv_types import JointDistributionHandmade
 from sympy.stats.rv import (RandomIndexedSymbol, random_symbols, RandomSymbol,
                             _symbol_converter, _value_check, pspace, given,
-                           dependent, is_random, sample_iter)
+                           dependent, is_random, sample_iter, Distribution)
 from sympy.stats.stochastic_process import StochasticPSpace
 from sympy.stats.symbolic_probability import Probability, Expectation
 from sympy.stats.frv_types import Bernoulli, BernoulliDistribution, FiniteRV
@@ -171,7 +171,7 @@ def state_space(self) -> tUnion[FiniteSet, Range]:
         return self.args[1]
 
     def distribution(self, key):
-        return None
+        return Distribution()
 
     def __call__(self, time):
         """
```
```
commit 72cf12b7af59bd0074075381a18538773370dfdd (HEAD -> issue_20078)
Author: Sidharth Mundhra <sidharthmundhra16@gmail.com>
Date:   Fri Dec 11 16:23:43 2020 +0530

    aims to fix simplification bug (#20078) with Wiener Process

commit ca83a68dd0b5e46a4e0d88ae9c1542f07affe782
Merge: e5c901e5f3 b1782d68f1
Author: Oscar Benjamin <oscar.j.benjamin@gmail.com>
Date:   Fri Dec 11 09:23:48 2020 +0000

    Merge pull request #19777 from theHamsta/__iter__
    
    Replace hasattr(... '__iter__') by check with `iterable`

commit e5c901e5f30bcc9bd0cb91f03d0696815b96bd1a
Merge: 87d967e81d 1fa5396d1e
Author: Oscar Benjamin <oscar.j.benjamin@gmail.com>
```
```python
In [1]: from sympy.stats import *

In [2]: T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])

In [3]: Y = DiscreteMarkovChain("Y", [0, 1, 2], T)

In [4]: simplify(Y[2]**2 + Y[2]**3)
Out[4]: 
               2
(Y[2] + 1)⋅Y[2] 
```

0sidharth committed:
Added abstract Distribution class in Stochastic Processes, and deprecation warnings for distribution

0sidharth commented:
Unfortunately I am having some troubles with github, messed up somewhere while rebasing it seems. Trying to fix.

czgdp1807 commented:
> Unfortunately I am having some troubles with github, messed up somewhere while rebasing it seems. Trying to fix.

What was the problem? Could you please provide details?

0sidharth commented:
I had done a soft reset to a previous commit (72cf12b7af59bd0074075381a18538773370dfdd) to add the `Distribution` class you had mentioned. After trying to commit I realised that this wouldn't work, so I went back to my then-latest commit, and did `git revert`, but it was still showing a revert commit message in log which I did not want, so I tried rebasing and squashing, and while I got the desired result locally, when I force-pushed this branch it was showing 15 commits to be merged instead of 2. 

Probably the error happened while squashing, and I later realised this could have much easily been done through hard resetting I guess :sweat_smile: . I think I might have fixed the issue by using `git reflog` to go back to previous state, but I did not want to spam the notifications of people subscribed to this PR by doing trial-and-error,  so I decided to move to a new PR.

czgdp1807 commented:
For me, [interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Changing-Multiple-Commit-Messages) always works and provides good number of options to drop commits from the history, squashing and re-wording.
